### DISCUSSION TITLE
How to add Depends for the docs page (openapi.json)?

### DISCUSSION THREAD

**Discussion by 20Koen02** [fastapi/fastapi#7103] [Questions] [answered]
### First check

* [*] I added a very descriptive title to this issue.
* [*] I used the GitHub search to find a similar issue and didn't find it.
* [*] I searched the FastAPI documentation, with the integrated search.
* [*] I already searched in Google "How to X in FastAPI" and didn't find any information.
* [*] I already read and followed all the tutorial in the docs and didn't find an answer.
* [*] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
* [*] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
* [*] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).
* [*] After submitting this, I commit to one of:
    * Read open issues with questions until I find 2 issues where I can help someone and add a comment to help there.
    * I already hit the "watch" button in this repository to receive notifications and I commit to help at least 2 people that ask questions in the future.
    * Implement a Pull Request for a confirmed bug.

### Description

Is there a way to add a dependency to the openapi.json endpoint? I want to implement an IP whitelist for the docs, so only certain people can view the documentation?
Only way i know how to do this is via a nginx configuration, but I want to do implement the whitelist in the application.

### Environment

* OS: [Windows 10]:
* FastAPI Version [0.61.1]:

**Comment by ycd** [comment]
Do you want to add a whitelisting to the `openapi.json` or to `/docs` or both?

**Comment by 20Koen02** [comment]
@ycd `/openapi.json`, as that will block all access to the docs when not whitelisted.
My plan is to have an ip attribute in my user model, and if the boolean attribute admin is true it will whitelist the associated ip. I can easily query these ip's to a list, but I don't know any ways to use this list as a whitelist for the `/openapi.json` endpoint.

**Comment by ycd** [comment]
Well you can override the `/openapi.json` path and add a basic check.


```python
from fastapi import FastAPI, Request, HTTPException

from fastapi.responses import JSONResponse
from fastapi.openapi.utils import get_openapi


WHLITELISTED_IPS = ["127.0.0.1"]


app = FastAPI(openapi_url=None)


def check_ip(request):
    if request.scope["client"][0] in WHLITELISTED_IPS:
        return True
    return False


@app.get("/openapi.json")
async def get_openapi_json(request: Request):
    if check_ip(request):
        return JSONResponse(get_openapi(title="FastAPI", version=1, routes=app.routes))
    raise HTTPException(status_code=401, detail="Unauthorized request.")
```

**Comment by Kludex** [comment]
Maybe:
`return JSONResponse(get_openapi(title="FastAPI", version=1, routes=app.routes))` -> `return JSONResponse(app.openapi_schema)`

**Comment by 20Koen02** [comment]
@ycd Thanks for your help, but this does not work. When `openapi_url=None`, fastapi skips a lot of steps, like the creation of the docs and redoc page.

@Kludex app.openapi_schema returns None for some reason. I used the same code fastapi uses internally in the code snippet below.

I have figured out a way how to do it, but it's very very hacky: We let fastapi create the openapi endpoint, but after fastapi's initialization, we remove the route and replace it with our own. I agree this is bad code practice, but its a valid workaround in my opinion.

```python
from fastapi import FastAPI, Request, HTTPException

from fastapi.responses import JSONResponse
from fastapi.openapi.utils import get_openapi

WHLITELISTED_IPS = ["127.0.0.1"]

app = FastAPI()

# All the routes are stored in app.routes
for route in app.routes:
    # Check if route is an instance of the class Route and not of any other class like BaseRoute
    # AND check if the name of the route is openapi
    if route.__class__ == Route and route.name == 'openapi':
        # Remove the route from the list
        app.router.routes.remove(route)


def check_ip(request):
    if request.scope["client"][0] in WHLITELISTED_IPS:
        return True
    return False

# Create a new openapi route, after the default one has been removed
@app.get("/openapi.json")
async def get_openapi_json(request: Request):
    if check_ip(request):
        return JSONResponse(get_openapi(
                title=app.title,
                version=app.version,
                openapi_version=app.openapi_version,
                description=app.description,
                routes=app.routes,
                tags=app.openapi_tags,
                servers=app.servers,
            ))
    raise HTTPException(status_code=401, detail="Unauthorized request.")
```

**Comment by ycd** [comment, marked as answer]
> When openapi_url=None, fastapi skips a lot of steps, like the creation of the docs and redoc page.

Expected, we create Redoc and Swagger using the `openapi_url` variable which is None, then we are using that variable inside `get_redoc_html` and `get_swagger_ui_html`. 

I'd still prefer this.


```python
from fastapi.openapi.docs import get_swagger_ui_html, get_redoc_html

@app.get("/docs")
async def get_swagger_documentation():
    return get_swagger_ui_html(openapi_url="/openapi.json", title="docs")


@app.get("/redoc")
async def get_redoc_documentation():
    return get_redoc_html(openapi_url="/openapi.json", title="docs")
```

**Comment by 20Koen02** [comment]
@ycd Sure, but there's a lot more code in the fastapi initialization, like the oauth redirection
For now I like my solution better because i don't feel like copying all the code from the FastAPI class.
Both solutions will work, but yours requires a lot more code to work for me.

**Comment by tiangolo** [comment]
Thanks for the help here everyone! :clap: :bow: 

Thanks for reporting back and closing the issue :+1:
